### PR TITLE
DEV: enable cors to all cdn get requests from workbox.

### PR DESCRIPTION
**Pull request by vinothkannans**
To prevent opaque cache files, now all the CDN files will be requested in 'cors' mode if the `cdn_cors_enabled` global setting is enabled. Before enabling the setting, should enable the cors in the CDN server by adding the response header `access-control-allow-origin: *` or `access-control-allow-origin: https://discourse.example.com`.

And other external file requests other than CDN will not be cached if the response type is opaque.

### CODE REVIEW

File: app/assets/javascripts/service-worker.js.erb [hunk 1]
```diff
@@ -4,24 +4,108 @@ importScripts("<%= "#{Discourse.asset_host}#{Discourse.base_path}/javascripts/wo
...
+if (cdnUrls.length > 0) {
+  var cdnCacheName = "cdn-" + cacheVersion;
+
+  var appendQueryStringPlugin = {
+    requestWillFetch: function (args) {
+      var request = args.request;
+
+      if (request.url.includes("avatar") || request.url.includes("emoji")) {
+        var url = new URL(request.url);
+        url.searchParams.append('refresh', 'true');
```
Comment (davidtaylorhq) [inline app/assets/javascripts/service-worker.js.erb:51]: I see we have a `cacheVersion` variable already in this file. Could we use that in the searchParams, so that it is easier to update in future?

File: app/assets/javascripts/service-worker.js.erb [hunk 2]
```diff
@@ -4,24 +4,108 @@ importScripts("<%= "#{Discourse.asset_host}#{Discourse.base_path}/javascripts/wo
...
+        maxEntries: 250,
+        purgeOnQuotaError: true, // safe to automatically delete if exceeding the available storage
+      }),
+    ],
+  })
+);
+
+var cdnUrls = [];
+
+<% if GlobalSetting.try(:cdn_cors_enabled) %>
```
Comment (davidtaylorhq) [inline app/assets/javascripts/service-worker.js.erb:39]: What's the plan for this global setting long-term? Will we remove it once the feature has been tested successfully across a few sites?

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -655,6 +656,18 @@ def check_xhr
...
   end
 
+  def block_cdn_requests
+    raise Discourse::NotFound if Discourse.is_cdn_request?(request.env)
+  end
+
+  def add_cors_header
+    if Discourse.is_cdn_request?(request.env) && ["GET", "OPTIONS"].include?(request.method)
+      response.headers['Access-Control-Allow-Origin'] = '*'
+      response.headers['Access-Control-Allow-Credentials'] = 'false'
```
Comment (davidtaylorhq) [inline app/controllers/application_controller.rb:666]: I don't think we need this: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials

> If you don't need credentials, omit this header entirely (rather than setting its value to false).

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -655,6 +656,18 @@ def check_xhr
...
+      response.headers['Access-Control-Allow-Methods'] = 'GET, OPTIONS'
```
Comment (davidtaylorhq) [inline app/controllers/application_controller.rb:667]: I wonder if we should add `HEAD` here as well?

File: config/initializers/008-rack-cors.rb [hunk 5]
```diff
@@ -25,15 +25,19 @@ def call(env)
...
+    Discourse::Cors.apply_headers(cors_origins, env, headers)
 
     [status, headers, body]
   end
 
   def self.apply_headers(cors_origins, env, headers)
-    origin = nil
+    if env['SCRIPT_NAME'] == "/assets" && ['GET', 'OPTIONS'].include?(env['REQUEST_METHOD']) && Discourse.is_cdn_request?(env)
+      headers['Access-Control-Allow-Origin'] = '*'
+      headers['Access-Control-Allow-Credentials'] = 'false'
```
Comment (davidtaylorhq) [inline config/initializers/008-rack-cors.rb:36]: As above

> If you don't need credentials, omit this header entirely (rather than setting its value to false).

Also... why are we doing this in two places? Could we do it all in the middleware, rather in both rails and middleware?

File: app/controllers/static_controller.rb [hunk 6]
```diff
@@ -6,6 +6,9 @@ class StaticController < ApplicationController
   skip_before_action :verify_authenticity_token, only: [:brotli_asset, :cdn_asset, :enter, :favicon, :service_worker_asset]
   skip_before_action :preload_json, only: [:brotli_asset, :cdn_asset, :enter, :favicon, :service_worker_asset]
   skip_before_action :handle_theme, only: [:brotli_asset, :cdn_asset, :enter, :favicon, :service_worker_asset]
+  skip_before_action :block_cdn_requests, only: [:brotli_asset, :cdn_asset, :enter, :favicon, :service_worker_asset]
+
+  before_action :add_cors_header, only: [:brotli_asset, :cdn_asset, :enter, :favicon, :service_worker_asset]
```
Comment (davidtaylorhq) [inline app/controllers/static_controller.rb:11]: I wonder if we could simplify this into an ApplicationController helper? So that we end up with something like

```
cdn_route :brotli_asset, :cdn_asset, :enter, :favicon, :service_worker_asset
```
And then cdn_route would be defined like
```
def cdn_route(*args)
  skip_before_action :block_cdn_requests, only: args
  before_action :add_cors_header, only: args
end
```

File: app/assets/javascripts/service-worker.js.erb (hunk 2 above)
Comment (vinothkannans) [reply app/assets/javascripts/service-worker.js.erb:39]: We're going to add this as an optional feature for a long time. Even after successfully tested in few sites.

File: app/assets/javascripts/service-worker.js.erb (hunk 1 above)
Comment (vinothkannans) [reply app/assets/javascripts/service-worker.js.erb:51]: I'm not sure here. Because we're not going to use this `searchParams` method for a long time. We will remove it once the cached files are refreshed (maybe in a year).

File: app/assets/javascripts/service-worker.js.erb (hunk 2 above)
Comment (davidtaylorhq) [reply app/assets/javascripts/service-worker.js.erb:39]: Why does it need to be optional? Is there some situation where it won't work?

Comment (vinothkannans) [reply app/assets/javascripts/service-worker.js.erb:39]: I think it won't raise any issues. Still, we should make sure that it won't make any problems/confusion for self-hosters. One problem I can imagine is it won't work properly if `DISCOURSE_CDN_ORIGIN_HOSTNAME` is required and not assigned.

**Comment by SamSaffron** [comment]
@davidtaylorhq / @vinothkannans  is this good for a merge?

**Comment by vinothkannans** [comment]
Yes, it should be good for the merge. @davidtaylorhq?

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -77,7 +77,7 @@ def script_asset_path(script)
         path = path.gsub(/\.([^.]+)$/, '.gz.\1')
       end
 
-    elsif GlobalSetting.cdn_url&.start_with?("https") && is_brotli_req?
+    elsif GlobalSetting.cdn_url&.start_with?("https") && is_brotli_req? && Rails.env != "development"
```
Comment (davidtaylorhq) [inline app/helpers/application_helper.rb:80]: Why is this change needed?

**Review by davidtaylorhq — APPROVED** [review]
Looks good to me, just a few questions about comments. 

Has this been tested on a production discourse instance? Because I am wondering how this will work for assets which are [served by `nginx` without hitting the rails app](https://github.com/discourse/discourse/blob/master/config/nginx.sample.conf#L255). (e.g. `/javascripts/*`)

**Comment by vinothkannans** [comment]
> Has this been tested on a production discourse instance?

Yes, dev is already running in this branch for a long time.

File: Gemfile.lock [hunk 8]
```diff
@@ -281,7 +281,7 @@ GEM
     rails_failover (0.5.6)
       activerecord (~> 6.0)
       railties (~> 6.0)
-    rails_multisite (2.3.0)
+    rails_multisite (2.5.0)
```
[the remark this answers is not in the document]
Comment (vinothkannans) [reply Gemfile.lock:284]: I think it's a bug related to GitHub. I had the same issue in my other PR.

File: app/assets/javascripts/service-worker.js.erb [hunk 9]
```diff
@@ -4,24 +4,108 @@ importScripts("<%= "#{Discourse.asset_host}#{Discourse.base_path}/javascripts/wo
...
-    return !(args.url.origin === location.origin && args.url.pathname.startsWith(authUrl));
+    return args.url.origin === location.origin && !args.url.pathname.startsWith(authUrl);
   }, // Match all except auth routes
   new workbox.strategies.NetworkFirst({ // This will only use the cache when a network request fails
-    cacheName: "discourse-" + cacheVersion,
+    cacheName: discourseCacheName,
     plugins: [
+      new workbox.cacheableResponse.Plugin({
+        statuses: [200] // opaque responses will return status code '0'
+      }), // for s3 secure media signed urls
```
[the remark this answers is not in the document]
Comment (vinothkannans) [reply app/assets/javascripts/service-worker.js.erb:27]: Because we're not adding the CORS header for S3 signed URLs. And I'm unable to find a way to do it.

File: app/helpers/application_helper.rb (hunk 7 above)
Comment (vinothkannans) [reply app/helpers/application_helper.rb:80]: In the development environment, we're not using/generating brotli assets. So we must skip this step in the dev environment.

**Comment by davidtaylorhq** [comment]
> dev is already running in this branch for a long time

Great! I think this is good to merge then - we can do more testing as it rolls out onto more sites via `DISCOURSE_CDN_CORS_ENABLED`. :+1:
